[demultiplexer] add a no-aggregation pipeline in DogStatsD

remeh opened the pull request:
### What does this PR do?

This PR adds a no-aggregation pipeline in the DogStatsD server.

The DogStatsD clients can send pre-aggregated metrics with their timestamp, these metrics will be forwarded to the intake without more processing except from tags enrichment.

### Motivation


Some DogStatsD clients or wrappers out there are already aggregating the samples during many seconds. They regularly send this data to the DogStatsD server but it means that it is receiving, in a short span of time, a lot of metrics ready to be sent to the intake. However, current pipeline is still 1. processing a context for every metric, 2. waiting for a few seconds for the time sampling, and 3. finally flushing it to the intake, which is inefficient since (1) & (2) don't need to be done for these metrics.

Having this no aggregation pipeline avoid all this extra work done for nothing.

### Additional Notes

* Everything is behind the feature flag `dogstatsd_no_aggregation_pipeline`, which is disabled by default.
* PR should be reviewable commit per commit, even if some things were already changed/removed between two commits.
* In the unit tests, a following PR will change the `WaitForSamples` signature to return both kind of samples, I did not do it in this PR to avoid confusing the reviewers with a bigger diff than necessary
 
### Possible Drawbacks / Trade-offs

This should have no impact on existing DogStatsD pipelines and behavior.

### Describe how to test/QA your changes


* *With feature disabled* (default): make sure usual DogStatsD traffic is working as usual
* *With feature enabled* (set `dogstatsd_no_aggregation_pipeline: true` or `DD_DOGSTATSD_NO_AGGREGATION_PIPELINE=true`):
  * make sure usual DogStatsD traffic is working as usual
  * send a serie with a timestamp 2 hours in the past through UDP using your shell: `echo -n "custom_metric:60|g|#env:test,dev:remy|T$(( $(date +%s)  -  7200 ))" | nc -4u -w0 127.0.0.1 8125`
  * validate that your metrics is visible at the correct time in a dashboard or in the metrics explorer

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

remeh committed:
[demultiplexer] buffer and flush historical metrics

remeh committed:
[dogstatsd] read the timestamp from the message and transport it in the "late sample"

remeh committed:
[demultiplexer] option for the no-agg pipeline + a few unit tests.

remeh committed:
[demultiplexer] rename `DemultiplexerOptions` -> `AgentDemultiplexerOptions`.

remeh committed:
[dogstatsd] improve metrics parsing unit test.

remeh committed:
[dogstatsd] polishing comments and telemetry.

remeh committed:
[demultiplexer] separate the flush mechanism of the no aggregation pipeline in its own worker.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/no-agg-pipeline

remeh committed:
[demultiplexer] the no aggregation worker is directly streaming metrics to the serializer.

remeh committed:
[demultiplexer] apply more polish

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/no-agg-pipeline

paulcacheux reviewed (approved):
For agent security owned files

gh123man commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,215 @@
+
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+	lastStream := time.Now().Add(time.Hour * 24)
```
What is the significance of this being 24 hours?

remeh replied:
A far enough date to not trigger an "automatic" flush too fast if one has just been done _and no metrics have been received meanwhile_. (that's the "and" which is important here)
So no significance, I'll add a comment on this 🙇

gh123man commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,215 @@
+				// the flush trigger may have set this flushBlockChan to a channel on which
+				// we need to send a signal to indicate the end of the flush.
+				if flushBlockChan != nil {
+					flushBlockChan <- struct{}{}
```
Would it be better to `close` the channel here instead of sending a signal and setting it to `nil`?

remeh replied:
I still have to set it to nil but you're right, would be better to close it I think.

gh123man reviewed (approved):
Mostly NITs and a couple questions - but LGTM. 
Should probably get a 2nd set of eyes from someone that knows this code better than I do :)

maxday reviewed (approved):
✅ for serverless, all integration tests are still passing 👍

AlexandreYang reviewed (approved):
Good for snmp related files

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/no-agg-pipeline

remeh commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,215 @@
+//
+// While streaming metrics to the serializer, the serializer should be responsible of sending the payloads
+// to the forwarder once one is generated (because full), even if it is still receiving metrics.
+// It's not how it works today, it will pile up the payloads, wait until the streaming is ending and then
```
Should be better

remeh commented on pkg/aggregator/demultiplexer_agent.go:
```diff
@@ -264,6 +320,11 @@ func (d *AgentDemultiplexer) Run() {
 
 	go d.aggregator.run()
+
+	if d.options.EnableNoAggregationPipeline {
```
Good catch! (otot of consistency that's safer)

remeh committed:
[demultiplexer] address feedback.

remeh committed:
[demultiplexer] in no-agg traffic, use same `interval` as usual dogstatsd traffic

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/no-agg-pipeline

remeh committed:
[demultiplexer] use a separate serializer for the no-agg pipeline

In order to avoid having both the regular pipelines and the no-agg one competing
with the series streaming.

ogaca-dd commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,218 @@
+						}
+
+					// flush signal
+					case trigger := <-w.flushChan:
```
You may not need a flush channel here or I miss something.

This part of the code is the [producer](https://github.com/DataDog/datadog-agent/blob/7.38.0/pkg/metrics/iterable_metrics.go#L117). When the producer finishes, `sendIterableSeries(w.serializer, start, serieSource)` returns, `serieConsumer` returns and `Serialize` returns.

When the Agent needs to flush, calling `return` should be enough.

remeh replied:
You miss a couple things, the flush channel is needed in this routine for two reasons:
  * this is the pattern we always use to avoid race conditions, any routine can call `flush()` and this mainloop routine will do it when it is not doing something else (such as receiving samples for instance)
  * it is indeed returning with the "break mainloop", I prefer this to a return in this case since it could confuse the reader, we're not returning for the main `run()` call but only this `Serializer` embedded lambda (btw this pattern in the serializer is once again really complex to use).

remeh replied:
Discussed offline: could be removed since returning from `Serialize` is enough in every case.

ogaca-dd commented on pkg/dogstatsd/parse.go:
```diff
@@ -139,20 +143,32 @@ func (p *parser) parseMetricSample(message []byte) (dogstatsdMetricSample, error
+		case bytes.HasPrefix(optionalField, timestampFieldPrefix):
+			if ts, err := strconv.ParseInt(string(optionalField[len(timestampFieldPrefix):]), 10, 0); err == nil {
+				timestamp = time.Unix(ts, 0)
+			}
```
When `err != nil` why not returning an error like for an invalid sample rate?

remeh replied:
True. I'll change this.

ogaca-dd commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,218 @@
+//   * listens for a flush signal
+// This is not ideal since the serializer should automatically takes the decision when to flush payloads to
+// the serializer but that's not how it works today, see noAggregationStreamWorker comment.
+func (w *noAggregationStreamWorker) run() {
```
It probably worths it to split this function.

ogaca-dd reviewed (changes requested):
Great PR!!!!

remeh commented on pkg/aggregator/demultiplexer_agent.go:
```diff
@@ -498,6 +504,14 @@ func (d *AgentDemultiplexer) GetEventsAndServiceChecksChannels() (chan []*metric
 // AddLateMetrics buffers a bunch of late metrics. This data will be directly
 // transmitted "as-is" (i.e. no sampling) to the serializer when a flush is triggered.
 func (d *AgentDemultiplexer) AddLateMetrics(samples metrics.MetricSampleBatch) {
+	// safe-guard: if for some reasons we are receiving some metrics here despite
```
The reason is that if a previous version of the Agent is receiving a metric with a timestamp, it will just ignore this timestamp field and process the metric as usual.
I think it's better to keep this behavior when this field is supported but the feature is disabled.

remeh committed:
[demultiplexer] address feedback review.

remeh commented on pkg/aggregator/demultiplexer_serverless.go:
```diff
@@ -147,11 +147,16 @@ func (d *ServerlessDemultiplexer) AddTimeSampleBatch(shard TimeSamplerID, sample
 
+// AddLateMetrics is not supported in the Serverless Agent implementation.
+func (d *ServerlessDemultiplexer) AddLateMetrics(samples metrics.MetricSampleBatch) {
+	panic("not implemented.")
```
I agree, but this is a feature that they are interested to implement in the future hence it's presence this way for now.
I'll move the one above tho (`AddCheckSample` to the `DemultiplexerWithAggregator` to avoid having this panic L152, because indeed here it doesn't make sense).

remeh commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,218 @@
+					// ticker regularly producing a flush signal if necessary
+					case <-ticker.C:
+						n := time.Now()
+						if serializedSamples > 0 && lastStream.Before(n.Add(-time.Second*1)) {
```
Your example is exactly the same but with a reversed test and moreover, a different meaning: I prefer to have the meaning related to the "last stream", because the feature is "we want to flush after x seconds if the stream has stopped" contrary to "we want to flush every x second at least".

remeh committed:
[demux] linter compliance.

remeh committed:
[demultiplexer] simplify the no-agg pipeline mainloop.

Removed the flush API: not needed since all flush must be triggered from the mainloop
Simplified the lastStream usage.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/no-agg-pipeline

ogaca-dd commented on pkg/aggregator/no_aggregation_stream_worker.go:
```diff
@@ -0,0 +1,218 @@
+				// we need to send a signal to indicate the end of the flush.
+				if flushBlockChan != nil {
+					close(flushBlockChan)
+					flushBlockChan = nil
```
Note: This code was removed.
